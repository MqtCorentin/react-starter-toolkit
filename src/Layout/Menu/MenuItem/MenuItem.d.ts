import { createMenu } from './MenuItem.container';

export type TActionElt = {
  url?: string;
  label: string;
};

// le container reçoit les items simples
export type TMenuItemBase = TActionElt & {
  children?: TMenuItemBase[];
  classModifier?: string;
  createMenuFn?: typeof createMenu;
};

// la vue reçoit les items avec des props fournies par le container
export type TMenuItem = TMenuItemBase & {
  ariaHaspopup?: string;
  ariaExpanded?: string;
  ariaLabel?: string;
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
};

export type TMenuItemEnhanced = TMenuItemBase & {
  authorized?: string[];
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
};
