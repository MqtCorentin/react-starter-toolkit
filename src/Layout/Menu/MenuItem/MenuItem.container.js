import { func } from 'prop-types';
import Authorize from 'shared/components/Authorize';
import { MenuItem } from './MenuItem';

export const createMenu = menuItems => menuItems.map(menuItem => <MenuItemEnhanced key={menuItem.label} {...menuItem} />);

const MenuItemEnhanced = ({ url, label, children, classModifier, authorized, hasFocus, index, tabIndex, createMenuFn }) => {
  const newClassModifier = [classModifier, children ? 'haschild' : ''].join(' ');
  const ariaProps = url ? { ariaHaspopup: 'true', ariaExpanded: 'false', ariaLabel: label } : {};

  return (
    <Authorize authorized={authorized}>
      <MenuItem
        hasFocus={hasFocus}
        index={index}
        tabIndex={tabIndex}
        createMenuFn={createMenuFn}
        url={url}
        label={label}
        classModifier={newClassModifier}
        {...ariaProps}
      >
        {children}
      </MenuItem>
    </Authorize>
  );
};

export const menuEnhancedPropTypes = {
  createMenuFn: func,
};

export const menuEnhancedDefaultProps = {
  createMenuFn: createMenu,
};

MenuItemEnhanced.propTypes = menuEnhancedPropTypes;
MenuItemEnhanced.defaultProps = menuEnhancedDefaultProps;

export default MenuItemEnhanced;
