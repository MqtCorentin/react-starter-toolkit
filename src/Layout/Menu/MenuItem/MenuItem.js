import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { NavBarItem } from '@axa-fr/react-toolkit-all';

export const ActionElt = ({ url, label }) =>
  url ? (
    <Link className="af-nav__link" to={url}>
      {label}
    </Link>
  ) : (
    <span className="af-nav__link">{label}</span>
  );

export const MenuItem = ({ classModifier, className, children, url, label, ariaHaspopup, ariaExpanded, ariaLabel, createMenuFn, ...rest }) => (
  <NavBarItem
    className={className}
    classModifier={classModifier}
    actionElt={ActionElt({ url, label })}
    aria-haspopup={ariaHaspopup}
    aria-expanded={ariaExpanded}
    aria-label={ariaLabel}
    {...rest}
  >
    {children && createMenuFn(children)}
  </NavBarItem>
);

export const MenuItemPropTypes = {
  url: PropTypes.string,
  label: PropTypes.string,
};

export const MenuItemDefaultProps = {
  url: null,
  label: '',
};

MenuItem.propTypes = MenuItemPropTypes;
MenuItem.defaultProps = MenuItemDefaultProps;

export default MenuItem;
