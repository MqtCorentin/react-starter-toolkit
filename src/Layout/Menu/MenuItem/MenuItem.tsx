import { Link } from 'react-router-dom';
import { NavBarItem } from '@axa-fr/react-toolkit-all';
import type { TActionElt, TMenuItem } from './MenuItem.d';

export const ActionElt = ({ url, label }: TActionElt) =>
  url ? (
    <Link className="af-nav__link" to={url}>
      {label}
    </Link>
  ) : (
    <span className="af-nav__link">{label}</span>
  );

export const MenuItem = ({ children, url, label, ariaHaspopup, ariaExpanded, ariaLabel, createMenuFn, ...rest }: TMenuItem) => (
  <NavBarItem {...rest} actionElt={ActionElt({ url, label })} aria-haspopup={ariaHaspopup} aria-expanded={ariaExpanded} aria-label={ariaLabel}>
    {children && createMenuFn?.(children)}
  </NavBarItem>
);

export default MenuItem;
