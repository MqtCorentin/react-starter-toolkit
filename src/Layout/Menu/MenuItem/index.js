export { default, createMenu } from './MenuItem.container';
export { MenuItem, MenuItemPropTypes } from './MenuItem';
