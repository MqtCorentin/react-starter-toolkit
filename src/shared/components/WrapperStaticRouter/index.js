import { StaticRouter } from 'react-router-dom/server';

export default ({ context = {}, children }) => <StaticRouter context={context}>{children}</StaticRouter>;
