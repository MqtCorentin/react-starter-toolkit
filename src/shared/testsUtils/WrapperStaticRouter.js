import React from 'react';
import { StaticRouter } from 'react-router-dom/server';

export default ({ context = {}, children }) => <StaticRouter context={context}>{children}</StaticRouter>;
