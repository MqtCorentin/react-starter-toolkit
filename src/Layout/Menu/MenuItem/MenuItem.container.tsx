import Authorize from 'shared/components/Authorize';
import { MenuItem } from './MenuItem';
import type { TMenuItemEnhanced, TMenuItemBase } from './MenuItem.d';

export const createMenu = (menuItems: TMenuItemBase[], onClick?: (event: React.MouseEvent<HTMLElement>) => void) =>
  menuItems.map(menuItem => <MenuItemEnhanced key={menuItem.label} {...menuItem} onClick={onClick} />);

const MenuItemEnhanced = ({ authorized, url, label, children, classModifier, createMenuFn = createMenu, ...props }: TMenuItemEnhanced) => {
  const newClassModifier = [classModifier, children ? 'haschild' : ''].join(' ').trim();
  const ariaProps = url ? { ariaHaspopup: 'true', ariaExpanded: 'false', ariaLabel: label } : {};

  return (
    <Authorize authorized={authorized}>
      <MenuItem {...props} createMenuFn={createMenuFn} url={url} label={label} classModifier={newClassModifier} {...ariaProps}>
        {children}
      </MenuItem>
    </Authorize>
  );
};

export default MenuItemEnhanced;
