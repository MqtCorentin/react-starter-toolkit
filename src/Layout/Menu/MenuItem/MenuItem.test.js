import React from 'react';
import { render } from '@testing-library/react';
import WrapperStaticRouter from 'shared/components/WrapperStaticRouter';
import { MenuItem, ActionElt } from './MenuItem';

describe('<MenuItem/>', () => {
  it('Render <Menu/>', () => {
    const { asFragment } = render(<MenuItem />, { wrapper: WrapperStaticRouter });
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('<ActionElt/>', () => {
  it('Render <ActionElt/> with Link When url is defined', () => {
    const { asFragment } = render(<ActionElt url="/members" label="withlink" />, { wrapper: WrapperStaticRouter });
    expect(asFragment()).toMatchSnapshot();
  });
  it('Render <ActionElt/> with Link When url is defined', () => {
    const { asFragment } = render(<ActionElt label="nolink" />, { wrapper: WrapperStaticRouter });
    expect(asFragment()).toMatchSnapshot();
  });
});
