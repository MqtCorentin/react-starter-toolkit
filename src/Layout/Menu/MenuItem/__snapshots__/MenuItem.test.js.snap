// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionElt/> Render <ActionElt/> with Link When url is defined 1`] = `
<DocumentFragment>
  <a
    class="af-nav__link"
    href="/members"
  >
    withlink
  </a>
</DocumentFragment>
`;

exports[`<ActionElt/> Render <ActionElt/> with Link When url is defined 2`] = `
<DocumentFragment>
  <span
    class="af-nav__link"
  >
    nolink
  </span>
</DocumentFragment>
`;

exports[`<MenuItem/> Render <Menu/> 1`] = `
<DocumentFragment>
  <li
    class="af-nav__item "
    role="none"
  >
    <span
      aria-expanded="false"
      class="af-nav__link"
      role="menuitem"
      tabindex="-1"
    />
  </li>
</DocumentFragment>
`;
